torchaudio/prototype/rnnt_decoder.py

```diff
@@ -0,0 +1,342 @@
+            new_hypos.append(
+                Hypo(
+                    tokens=new_tokens,
+                    predictor_out=pred_out[i].detach(),
+                    state=_slice_state(pred_states, i, device),
+                    score=scores[i],
+                    ali=h_a.ali + [t],
+                    blank=self.blank,
+                    key=str(new_tokens),
+                )
+            )
+        return new_hypos
+
+    def _search(
+        self, enc_out: torch.Tensor, hypo: Optional[Hypo], beam_width: int,
+    ) -> List[Hypo]:
+        n_time_steps = enc_out.shape[1]
+        device = enc_out.device
+
+        a_hypos: List[Hypo] = []
+        b_hypos = self._init_b_hypos(hypo, device)
+        for t in range(n_time_steps):
+            a_hypos = b_hypos
+            b_hypos = torch.jit.annotate(List[Hypo], [])
+            key_to_b_hypo: Dict[str, Hypo] = {}
+            symbols_current_t = 0
+
+            while a_hypos:
+                next_token_probs = self._gen_next_token_probs(
+                    enc_out[:, t: t + 1], a_hypos, device
+                )
+                next_token_probs = next_token_probs.cpu()
+                b_hypos = self._gen_b_hypos(
+                    b_hypos, a_hypos, next_token_probs, key_to_b_hypo,
+                )
+
+                if symbols_current_t == self.step_max_tokens:
+                    break
+
+                a_hypos = self._gen_a_hypos(
+                    a_hypos, b_hypos, next_token_probs, t, beam_width, device,
+                )
+                if a_hypos:
+                    symbols_current_t += 1
+
+            _, sorted_idx = torch.tensor(
+                [self.hypo_sort_key(hypo) for hypo in b_hypos]
+            ).topk(beam_width)
+            b_hypos = [b_hypos[idx] for idx in sorted_idx]
+
+        return b_hypos
+
+    def forward(
+        self, input: torch.Tensor, length: torch.Tensor, beam_width: int
+    ) -> List[Hypo]:
+        r"""Performs beam search for the given input sequence.
+
+        T: number of frames;
+        D: feature dimension of each frame.
+
+        Args:
+            input (torch.Tensor): sequence of input frames, with shape (1, T, D).
+            length (torch.Tensor): number of valid frames in input
+                sequence, with shape (1,).
+            beam_width (int): beam size to use during search.
+
+        Returns:
+            List[Hypo]: top-``beam_width`` hypotheses found by beam search.
+        """
+        assert (
+            len(input.shape) == 3 and input.shape[0] == 1
+        ), "input must be of shape (1, T, D)"
+        assert length.shape == (1,), "length must be of shape (1,)"
+        enc_out, _ = self.model.transcribe(input, length)
+        return self._search(enc_out, None, beam_width)
+
+    @torch.jit.export
+    def infer(
+        self,
+        input: torch.Tensor,
+        length: torch.Tensor,
+        state: Optional[List[List[torch.Tensor]]],
+        hypo: Optional[Hypo],
```

Comment:
Without a default value of `= None`, isn't this mandatory argument?

Reply:
yeah it is — changed `state` and `hypo` to kwargs